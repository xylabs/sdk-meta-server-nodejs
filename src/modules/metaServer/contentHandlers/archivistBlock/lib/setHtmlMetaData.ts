import { BoundWitnessSchema } from '@xyo-network/boundwitness-model'
import { PayloadWrapper } from '@xyo-network/payload-wrapper'
import { Meta, metaBuilder } from '@xyo-network/sdk-meta'
import cloneDeep from 'lodash/cloneDeep'

import { PayloadInfo } from '../../../types'
import { getArchivistForDomain } from './getArchivistForDomain'

export const setHtmlMetaData = async (info: PayloadInfo, html: string, config: Meta): Promise<string> => {
  const { apiDomain, hash, path } = info
  const meta = cloneDeep(config)
  if (hash && apiDomain) {
    try {
      const archivist = await getArchivistForDomain(apiDomain)
      const results = await archivist.get([hash])
      const payload = PayloadWrapper.tryParse(results?.[0])
      if (payload) {
        const schema = payload.schema()
        const type = schema === BoundWitnessSchema ? 'Bound Witness' : 'Payload'
        meta.title = `XYO 2.1: ${type} | ${hash}`
        meta.description = `A XYO 2.1 ${schema} payload with the hash ${hash}.`
      }
    } catch (error) {
      console.log(error)
    }
  }
  meta.og = { ...meta.og, title: meta.title, url: path }
  meta.twitter = { ...meta.twitter, title: meta.title }
  const updated = metaBuilder(html, meta)
  return updated
}
