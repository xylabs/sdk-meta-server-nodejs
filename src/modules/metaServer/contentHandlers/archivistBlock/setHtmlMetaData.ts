import { XyoBoundWitnessSchema } from '@xyo-network/boundwitness-model'
import { PayloadWrapper } from '@xyo-network/payload-wrapper'
import { Meta, metaBuilder } from '@xyo-network/sdk-meta'
import cloneDeep from 'lodash/cloneDeep'

import { PayloadInfo } from '../../types'
import { getArchivistForDomain } from './getArchivistForDomain'

export const setHtmlMetaData = async (info: PayloadInfo, html: string, config: Meta): Promise<string> => {
  const { apiDomain, hash, path } = info
  const meta = cloneDeep(config)
  if (hash && apiDomain) {
    try {
      const archivist = await getArchivistForDomain(apiDomain)
      const results = await archivist.get([hash])
      const payload = PayloadWrapper.tryParse(results?.[0])
      if (payload) {
        const schema = payload.schema
        const type = schema === XyoBoundWitnessSchema ? 'Bound Witness' : 'Payload'
        meta.title = `XYO 2.0: ${type} | ${hash}`
        meta.description = `A XYO 2.0 ${schema} payload with the hash ${hash}.`
      }
    } catch (error) {
      console.log(error)
    }
  }
  meta.og = { ...meta.og, title: meta.title, url: path } as typeof meta.og
  meta.twitter = { ...meta.twitter, title: meta.title }
  const updated = metaBuilder(html, meta)
  return updated
}
