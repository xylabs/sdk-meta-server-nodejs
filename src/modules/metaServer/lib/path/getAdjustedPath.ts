import Path from 'node:path'

import { Request } from 'express'

import { isKnownFileExtension } from '../file/index.js'

export const getAdjustedPath = (req: Request): string => {
  if (isKnownFileExtension(req.path)) return req.path
  return Path.join(req.path, 'index.html')
}
