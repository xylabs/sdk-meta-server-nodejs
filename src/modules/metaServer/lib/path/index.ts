export * from './getAdjustedPath'
export * from './isHtmlLike'
