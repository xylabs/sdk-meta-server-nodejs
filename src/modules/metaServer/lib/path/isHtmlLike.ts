import Path from 'node:path'

import { Request } from 'express'

import { isKnownFileExtension } from '../file/index.js'

/**
 * Returns true request path could be referring
 * to an HTML file
 * @param req The web request
 * @returns True if the request path could be referring,
 * to an HTML document, false otherwise
 */
export const isHtmlLike = (req: Request): boolean => {
  const ext = Path.extname(req.path)
  if (ext.length === 0) return true
  if (ext.toLowerCase() === '.html') return true
  if (!isKnownFileExtension(req.path)) return true
  return false
}
