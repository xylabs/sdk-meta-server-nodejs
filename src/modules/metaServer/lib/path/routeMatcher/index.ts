export * from './createGlobMatcher.js'
export * from './createRegexMatcher.js'
export * from './RouteMatcher.js'
