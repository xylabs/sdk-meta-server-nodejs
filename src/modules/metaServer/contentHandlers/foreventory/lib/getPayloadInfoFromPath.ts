import { getArchivistDomainFromUri, getHashInfoFromUri } from '../../../lib'
import { PayloadInfo } from '../../../types'

export const getPayloadInfoFromPath = (path: string): PayloadInfo => {
  const apiDomain = getArchivistDomainFromUri(path)
  const { hash } = getHashInfoFromUri(path)
  return { apiDomain, hash, path }
}
