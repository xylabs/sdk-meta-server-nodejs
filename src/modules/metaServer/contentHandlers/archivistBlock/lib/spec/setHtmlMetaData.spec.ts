import { Meta, OpenGraphMeta, TwitterMeta } from '@xyo-network/sdk-meta'
import { readFile } from 'fs/promises'
import { join } from 'path'

import { getPayloadInfoFromPath } from '../getPayloadInfoFromPath'
import { setHtmlMetaData } from '../setHtmlMetaData'

const title = 'New Title'
const testOgMeta: OpenGraphMeta = {
  image: {
    secure_url: 'https://www.fillmurray.com/200/300',
    url: 'https://www.fillmurray.com/200/300',
  },
  title,
  type: 'website',
  url: 'https://test.xyo.network',
}
const testTwitterMeta: TwitterMeta = {
  card: 'summary_large_image',
  image: { '': 'https://www.placecage.com/400/300' },
  title,
}
const testMeta: Meta = {
  description: 'New Description',
  og: testOgMeta,
  title,
  twitter: testTwitterMeta,
}

const expectedOgMeta: OpenGraphMeta = {
  image: {
    secure_url: 'https://explore.xyo.network/meta-image-explore-og.jpg',
    url: 'https://explore.xyo.network/meta-image-explore-og.jpg',
  },
  title: 'XYO 2.1 Explore',
  type: 'website',
}
const expectedTwitterMeta: TwitterMeta = {
  card: 'summary_large_image',
  image: { '': 'https://explore.xyo.network/meta-image-explore-twitter.jpg' },
  title: 'XYO 2.1 Explore',
}
const expectedMeta: Meta = {
  description: 'Explore the XYO 2.1 Blockchain',
  og: expectedOgMeta,
  twitter: expectedTwitterMeta,
}

const verifyHtmlContainsMeta = (html: string, path: string, meta = testMeta) => {
  expect(html).toContain(meta.title)
  expect(html).toContain(meta.description)
  expect(html).toContain((meta.og?.image as { url?: string })?.url)
  expect(html).toContain(meta.og?.title)
  expect(html).toContain(meta.og?.type)
  expect(html).toContain(path)
  expect(html).toContain(meta.twitter?.card)
  expect(html).toContain(meta.twitter?.title)
  expect(html).toContain(meta.twitter?.image?.[''])
}

describe('setHtmlMetaData', () => {
  let originalHtml: string
  beforeAll(async () => {
    const template = join(__dirname, 'template.html')
    originalHtml = await readFile(template, { encoding: 'utf-8' })
  })
  describe('for non-Payload URLs', () => {
    it('sets standard meta fields', async () => {
      const path = 'https://www.google.com'
      const info = getPayloadInfoFromPath(path)
      const newHtml = await setHtmlMetaData(info, originalHtml, testMeta)
      verifyHtmlContainsMeta(newHtml, path)
    })
  })
  describe('for Payload URLs sets standard meta fields', () => {
    const bwHash = '6d7351f818fd9342fc41072a7dcceb57bdff0fe55e2f8e4d28abac13a5320b15'
    const payloadHash = '9213d6b605aad8a4b1871fd5d4f9a23355ffb7334b290afae51ead9bda9fa6a3'
    const cases: [type: string, hash: string, path: string][] = [
      ['Bound Witness', bwHash, `https://beta.explore.xyo.network/payload/hash/${bwHash}?network=kerplunk`],
      ['Payload', payloadHash, `https://beta.explore.xyo.network/payload/hash/${payloadHash}?network=kerplunk`],
    ]
    it.only.each(cases)('sets fields based on the %s', async (type, hash, path) => {
      const info = getPayloadInfoFromPath(path)
      const newHtml = await setHtmlMetaData(info, originalHtml, expectedMeta)
      expect(newHtml).toMatchSnapshot()
    })
  })
})
