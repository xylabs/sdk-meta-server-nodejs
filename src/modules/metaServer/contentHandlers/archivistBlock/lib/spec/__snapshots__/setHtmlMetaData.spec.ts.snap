// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`setHtmlMetaData for Payload URLs sets standard meta fields sets fields based on the Bound Witness 1`] = `
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <link rel="icon"
          href="/favicon.ico"
    >
    <meta name="viewport"
          content="width=device-width,initial-scale=1"
    >
    <meta name="theme-color"
          content="#000000"
    >
    <meta name="description"
          content="Own your piece of XYO's Decentralized Digital World!"
    >
    <link rel="apple-touch-icon"
          href="/apple-touch-icon.png"
    >
    <link rel="manifest"
          href="/manifest.json"
    >
    <title>
      XYO 2.1: Bound Witness | 6d7351f818fd9342fc41072a7dcceb57bdff0fe55e2f8e4d28abac13a5320b15
    </title>
    <link href="https://fonts.googleapis.com/css?family=Nunito+Sans|Lexend+Deca|Rock+Salt|Source+Code+Pro&amp;display=swap"
          rel="stylesheet"
    >
    <meta property="og:url"
          content="https://beta.explore.xyo.network/payload/hash/6d7351f818fd9342fc41072a7dcceb57bdff0fe55e2f8e4d28abac13a5320b15?network=kerplunk"
    >
    <meta property="og:type"
          content="website"
    >
    <meta property="og:title"
          content="XYO 2.1: Bound Witness | 6d7351f818fd9342fc41072a7dcceb57bdff0fe55e2f8e4d28abac13a5320b15"
    >
    <meta property="og:description"
          content="Explore the XYO 2.1 Blockchain"
    >
    <meta property="og:image"
          content="https://explore.xyo.network/meta-image-explore.jpg"
    >
    <meta property="twitter:card"
          content="summary_large_image"
    >
    <meta property="twitter:image"
          content="https://explore.xyo.network/meta-image-explore-twitter.jpg"
    >
    <meta property="twitter:site"
          content="@OfficialXYO"
    >
    <meta property="twitter:creator"
          content="@OfficialXYO"
    >
    <script async
            src="https://www.googletagmanager.com/gtag/js?id=G-795QBPW744"
    >
    </script>
    <script>
      function gtag() { dataLayer.push(arguments) } window.dataLayer = window.dataLayer || [], gtag("js", new Date), gtag("config", "G-795QBPW744")
    </script>
    <style>
      html {
      overflow-y: auto;
      overflow-x: hidden
    }

    #root,
    body {
      height: 100%;
      width: 100%;
      margin: 0;
      padding: 0
    }
    </style>
    <script defer="defer"
            src="/static/js/main.ae7f7033.js"
    >
    </script>
    <link href="/static/css/main.026e3fe6.css"
          rel="stylesheet"
    >
    <meta property="description"
          content="A XYO 2.1 network.xyo.boundwitness payload with the hash 6d7351f818fd9342fc41072a7dcceb57bdff0fe55e2f8e4d28abac13a5320b15."
    >
    <meta property="og:image:secure_url"
          content="https://explore.xyo.network/meta-image-explore-og.jpg"
    >
    <meta property="og:image:url"
          content="https://explore.xyo.network/meta-image-explore-og.jpg"
    >
    <meta property="twitter:title"
          content="XYO 2.1: Bound Witness | 6d7351f818fd9342fc41072a7dcceb57bdff0fe55e2f8e4d28abac13a5320b15"
    >
    <meta property="title"
          content="XYO 2.1: Bound Witness | 6d7351f818fd9342fc41072a7dcceb57bdff0fe55e2f8e4d28abac13a5320b15"
    >
  </head>
  <body style="padding:0;margin:0;overflow-x:hidden">
    <noscript>
      <iframe src="https://www.googletagmanager.com/ns.html?id=GTM-W2TFNXL"
              height="0"
              width="0"
              style="display:none;visibility:hidden"
      >
      </iframe>
    </noscript>
    <noscript>
      You need to enable JavaScript to run this
    app.
    </noscript>
    <div id="root">
    </div>
  </body>
</html>
`;

exports[`setHtmlMetaData for Payload URLs sets standard meta fields sets fields based on the Payload 1`] = `
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <link rel="icon"
          href="/favicon.ico"
    >
    <meta name="viewport"
          content="width=device-width,initial-scale=1"
    >
    <meta name="theme-color"
          content="#000000"
    >
    <meta name="description"
          content="Own your piece of XYO's Decentralized Digital World!"
    >
    <link rel="apple-touch-icon"
          href="/apple-touch-icon.png"
    >
    <link rel="manifest"
          href="/manifest.json"
    >
    <title>
      XYO 2.1: Payload | 9213d6b605aad8a4b1871fd5d4f9a23355ffb7334b290afae51ead9bda9fa6a3
    </title>
    <link href="https://fonts.googleapis.com/css?family=Nunito+Sans|Lexend+Deca|Rock+Salt|Source+Code+Pro&amp;display=swap"
          rel="stylesheet"
    >
    <meta property="og:url"
          content="https://beta.explore.xyo.network/payload/hash/9213d6b605aad8a4b1871fd5d4f9a23355ffb7334b290afae51ead9bda9fa6a3?network=kerplunk"
    >
    <meta property="og:type"
          content="website"
    >
    <meta property="og:title"
          content="XYO 2.1: Payload | 9213d6b605aad8a4b1871fd5d4f9a23355ffb7334b290afae51ead9bda9fa6a3"
    >
    <meta property="og:description"
          content="Explore the XYO 2.1 Blockchain"
    >
    <meta property="og:image"
          content="https://explore.xyo.network/meta-image-explore.jpg"
    >
    <meta property="twitter:card"
          content="summary_large_image"
    >
    <meta property="twitter:image"
          content="https://explore.xyo.network/meta-image-explore-twitter.jpg"
    >
    <meta property="twitter:site"
          content="@OfficialXYO"
    >
    <meta property="twitter:creator"
          content="@OfficialXYO"
    >
    <script async
            src="https://www.googletagmanager.com/gtag/js?id=G-795QBPW744"
    >
    </script>
    <script>
      function gtag() { dataLayer.push(arguments) } window.dataLayer = window.dataLayer || [], gtag("js", new Date), gtag("config", "G-795QBPW744")
    </script>
    <style>
      html {
      overflow-y: auto;
      overflow-x: hidden
    }

    #root,
    body {
      height: 100%;
      width: 100%;
      margin: 0;
      padding: 0
    }
    </style>
    <script defer="defer"
            src="/static/js/main.ae7f7033.js"
    >
    </script>
    <link href="/static/css/main.026e3fe6.css"
          rel="stylesheet"
    >
    <meta property="description"
          content="A XYO 2.1 network.xyo.blockchain.ethereum.gas payload with the hash 9213d6b605aad8a4b1871fd5d4f9a23355ffb7334b290afae51ead9bda9fa6a3."
    >
    <meta property="og:image:secure_url"
          content="https://explore.xyo.network/meta-image-explore-og.jpg"
    >
    <meta property="og:image:url"
          content="https://explore.xyo.network/meta-image-explore-og.jpg"
    >
    <meta property="twitter:title"
          content="XYO 2.1: Payload | 9213d6b605aad8a4b1871fd5d4f9a23355ffb7334b290afae51ead9bda9fa6a3"
    >
    <meta property="title"
          content="XYO 2.1: Payload | 9213d6b605aad8a4b1871fd5d4f9a23355ffb7334b290afae51ead9bda9fa6a3"
    >
  </head>
  <body style="padding:0;margin:0;overflow-x:hidden">
    <noscript>
      <iframe src="https://www.googletagmanager.com/ns.html?id=GTM-W2TFNXL"
              height="0"
              width="0"
              style="display:none;visibility:hidden"
      >
      </iframe>
    </noscript>
    <noscript>
      You need to enable JavaScript to run this
    app.
    </noscript>
    <div id="root">
    </div>
  </body>
</html>
`;
